### PR TITLE
fix: Fix bug and add test case

### PR DESCRIPTION
**Pull request by p-offtermatt**
## Description

Closes: #2169

Bug reported by @fastfadingviolets via Slack

In the test scenario presented in the new e2e test, namely:
* provider chain with max_validators=5, max_provider_consensus_validators=4
* validators with powers 30, 29, 20, 10, 7, 4
* a consumer chain with top N=100
* then validators 30, 29, 20, 10 should have to validate on the consumer chain.

The previous behaviour was that validator 10 did not have to validate, and only 30, 29, and 20 had to validate.

The bug was this:
In https://github.com/cosmos/interchain-security/blob/0e080c3e4ad2df5667695d890740175d8212c4c4/x/ccv/types/utils.go#L134 in the function GetLastBondedValidatorsUtil, the staking keeper method IterateLastValidatorPowers is called assuming that it returns validators sorted by power.
This is not true, so wrong validators would be returned, which caused wrong validators to be used when computing the consumer chain validator set.

The fix is in x/ccv/types/utils.go

tests/e2e contains a regression tests outlining the scenario explained above.

In the integration tests, a test actually depended on the broken behaviour, so it was adjusted.

Further, the mocks were adjusted, since GetLastBondedValidatorsUtil calls GetBondedValidatorsByPower instead of IterateLastValidatorPowers now, and doesn't need to take the powers as an input.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] Added `!` to the type prefix if the change is [state-machine breaking](https://github.com/cosmos/interchain-security/blob/main/RELEASES.md#breaking-changes)
* [ ] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [ ] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/00-intro.md)
* [ ] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` the type prefix if the change is state-machine breaking
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

## Summary by CodeRabbit

- **New Features**
  - Introduced new test configurations for inactive validators, enhancing the testing framework's flexibility.
  - Added logging for validator opting-in processes, improving traceability.

- **Bug Fixes**
  - Enhanced assertions by sorting staked tokens in integration tests for more accurate validation.

- **Refactor**
  - Simplified function calls in mock setups by removing unnecessary parameters across various tests.
  - Refactored logic for retrieving last bonded validators, improving efficiency and error handling.

- **Tests**
  - Expanded test scenarios to include edge cases related to inactive validators and their configurations.

### CODE REVIEW

File: x/ccv/types/utils.go [hunk 1]
```diff
@@ -128,38 +129,18 @@ func GetConsAddrFromBech32(bech32str string) (sdk.ConsAddress, error) {
...
-	lastPowers := make([]stakingtypes.LastValidatorPower, maxVals)
-
-	i := 0
-	err := stakingKeeper.IterateLastValidatorPowers(ctx, func(addr sdk.ValAddress, power int64) (stop bool) {
-		lastPowers[i] = stakingtypes.LastValidatorPower{Address: addr.String(), Power: power}
-		i++
-		return i >= int(maxVals) // stop iteration if true
-	})
+	// get the bonded validators from the staking module, sorted by power
+	bondedValidators, err := stakingKeeper.GetBondedValidatorsByPower(ctx)
```
Comment (MSalopek) [inline x/ccv/types/utils.go:133]: This was super subtle. Great catch!

**Review by MSalopek — APPROVED** [review]
Approving the change, I have no objections to testing changes.

File: x/ccv/types/utils.go [hunk 2]
```diff
@@ -128,38 +129,18 @@ func GetConsAddrFromBech32(bech32str string) (sdk.ConsAddress, error) {
...
 	if err != nil {
-		return nil, err
+		panic(fmt.Errorf("failed to get bonded validators: %w", err))
```
Comment (mpoke) [inline x/ccv/types/utils.go:135]: Return error instead of panicking.

File: x/ccv/provider/keeper/partial_set_security.go [hunk 3]
```diff
@@ -91,6 +91,9 @@ func (k Keeper) HandleOptOut(ctx sdk.Context, chainID string, providerAddr types
 // OptInTopNValidators opts in to `chainID` all the `bondedValidators` that have at least `minPowerToOptIn` power
 func (k Keeper) OptInTopNValidators(ctx sdk.Context, chainID string, bondedValidators []stakingtypes.Validator, minPowerToOptIn int64) {
 	for _, val := range bondedValidators {
+		// log the validator
+		k.Logger(ctx).Info("Checking whether to opt in validator because of top N", "validator", val.GetOperator())
```
Comment (mpoke) [inline x/ccv/provider/keeper/partial_set_security.go:95]: Either change to debug or remove. This might lead to a lot of logging messages. One for every TopN validator.

File: x/ccv/provider/keeper/partial_set_security.go [hunk 4]
```diff
@@ -111,9 +114,11 @@ func (k Keeper) OptInTopNValidators(ctx sdk.Context, chainID string, bondedValid
 				continue
 			}
 
+			k.Logger(ctx).Info("Opting in validator", "validator", val.GetOperator())
```
Comment (mpoke) [inline x/ccv/provider/keeper/partial_set_security.go:117]: ditto

File: x/ccv/provider/keeper/partial_set_security.go (hunk 3 above)
Comment (p-offtermatt) [reply x/ccv/provider/keeper/partial_set_security.go:95]: Changing to debug!

File: x/ccv/types/utils.go (hunk 2 above)
Comment (p-offtermatt) [reply x/ccv/types/utils.go:135]: thanks, good catch

**Review by mpoke — APPROVED** [review]
Great work @p-offtermatt
